### ISSUE TITLE
docker restart fails

### ISSUE THREAD

**Issue by jordic** [docker/docker#22042]
**Output of `docker version`:**

```
Client:
 Version:      1.11.0
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   4dc5990
 Built:        Wed Apr 13 18:34:23 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.0
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   4dc5990
 Built:        Wed Apr 13 18:34:23 2016
 OS/Arch:      linux/amd64

```

**Output of `docker info`:**

```
Containers: 17
 Running: 9
 Paused: 0
 Stopped: 8
Images: 63
Server Version: 1.11.0
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 205
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null host bridge
Kernel Version: 3.13.0-24-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 490.1 MiB
Name: anileva.com
ID: TXF3:AQB2:HSIW:2TJJ:ZD2N:BQMQ:GGWT:V4TG:53Y6:ZYW7:4XBD:SVJK
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
WARNING: No kernel memory limit support
```

Ubuntu 14.04 on digital ocean.

**Steps to reproduce the issue:**
1. docker restart
2. container fails to restart with a message:

**Describe the results you received:**
This is the daemon log:
```
time="2016-04-14T12:32:10.825211877+02:00" level=info msg="Loading containers: done."
time="2016-04-14T12:32:10.825372308+02:00" level=info msg="Daemon has completed initialization"
time="2016-04-14T12:32:10.825529766+02:00" level=info msg="Docker daemon" commit=4dc5990 graphdriver=aufs version=1.11.0
time="2016-04-14T12:32:10.842463311+02:00" level=info msg="API listen on /var/run/docker.sock"
time="2016-04-14T12:57:08.140573669+02:00" level=error msg="restartmanager canceled"
time="2016-04-14T12:57:08.458286186+02:00" level=error msg="Handler for POST /v1.23/containers/artte/restart returned error: Cannot restart container artte: remove /var/run/docker/libcontainerd/ec46eb1b1f2dac498a316ae9615c3019f5d4c08e7b98c42e85741ebcd4467b31/rootfs: device or resource busy"
time="2016-04-14T12:57:19.072840744+02:00" level=error msg="Handler for POST /v1.23/containers/artte/stop returned error: Container artte is already stopped"
time="2016-04-14T12:57:23.758287281+02:00" level=error msg="Handler for POST /v1.23/containers/artte/start returned error: remove /var/run/docker/libcontainerd/ec46eb1b1f2dac498a316ae9615c3019f5d4c08e7b98c42e85741ebcd4467b31/rootfs: device or resource busy"
time="2016-04-14T12:57:31.497645216+02:00" level=info msg="Processing signal 'terminated'"
```

**Describe the results you expected:**
Container restarted

**Additional information you deem important (e.g. issue happens only occasionally):**

- The container is using some shared volumes on host. on 1.10, works well, but sometimes when I try to remove a container I receive a similar message..  but seems something related to aufs

- Container is launched with the restart=always param. 

**As a workaround I can restart the docker daemon, but is a no go.. **

**Comment by WeiZhang555** [comment]
I observed this before in https://github.com/docker/docker/pull/21854, it can be reproduced with this test case `TestRestartContainerwithRestartPolicy`,  see code in https://github.com/docker/docker/blob/master/integration-cli/docker_cli_restart_test.go#L231-L243 

I didn't find the root cause, but guess there must be some racey problems between stop and start, it's worth attention.

**Comment by tonistiigi** [comment]
> *As a workaround I can restart the docker daemon, but is a no go.. *

Does this mean that trying to start that container manually again also fails?

**Comment by jordic** [comment]
I think its not the same bug. I dont restart an already stopped container. I just try to restart a running container and fails to stop due resource busy... It happens all the time i can't restart the container without restarting the entry daemon.. :[

**Comment by jordic** [comment]
Yes @tonistiigi. The only way of restarting it is restarting the daemon..

**Comment by WeiZhang555** [comment]
Ah, so it's truly a different bug, but the error message is quite similar, maybe they are related to each other.

**Comment by jordic** [comment]
I discovered something else: I'm using cadvisor on the same host for metrics.
Seems that if I stop cadvisor before trying to restart a container, all works as expected. 
Can be something related to cadvisor, and the way it watches the daemon? (Or volumes created by the daemon?)
@thaJeztah @tonistiigi

**Comment by thaJeztah** [comment]
Sounds plausible, running cAdvisor is a known issue indeed, but wondering why you didn't run into this on 1.10.3

**Comment by jordic** [comment]
I have another similar issue (less agressive) .with 1.10.3 on some busy container can't rm them. And ended renaming and let them be gc on daemon restart, with a similar error message related to a busy volume. I still have some machines at 10.3 and will try to reproduce with and without cadvisor...

**Comment by crosbymichael** [comment]
This was fixed in #22256.

Thanks!
